In [1]:
import pandas as pd

In [2]:
import os

In [3]:
base_dir = os.path.join('data', 'bookmaker')

In [4]:
base_dir

'data/bookmaker'

In [5]:
log_filename = os.path.join(base_dir, 'log.csv')
print(log_filename)

data/bookmaker/log.csv


In [6]:
users_filename = os.path.join(base_dir, 'users.csv')
print(users_filename)

data/bookmaker/users.csv


In [7]:
sample_filename = os.path.join(base_dir, 'sample.csv')

#### Task 1. Используйте оригинальный датасет log.csv. Результат запишите числом в поле ниже. Посчитайте, сколько раз люди приходили, но не делали ставок. Для этого заполните пропуски в столбце bet значением 0 и посчитайте количество таких значений.
Подсказка: можно использовать value_counts().

#### Solution:

In [25]:
# прочитаем файл в переменную log
log = pd.read_csv(log_filename, header=None)

# добавим названия колонок user_id, time, bet, win
log.columns = ['user_id','time','bet','win']

# Заполним нулями пропущенные значения в колонке 'bet'
log['bet'].fillna(0, inplace=True)

# сосчитаем количества вхождений разных вариантов данных в колонке 'bet' и выведем на экран 
display(log['bet'].value_counts())

0.0          515
500.0         50
100.0         48
300.0         42
800.0         42
200.0         40
400.0         40
700.0         38
600.0         37
1000.0        37
900.0         30
9754.0        10
10554.0       10
10154.0        7
10254.0        7
10454.0        6
9954.0         6
10754.0        5
10654.0        5
10354.0        4
10054.0        4
9854.0         2
5000.0         1
27000.0        1
9876.0         1
156789.0       1
12945.0        1
12548.0        1
145732.0       1
1678321.0      1
98753.0        1
123981.0       1
104540.0       1
8700.0         1
950.0          1
8734.0         1
7650.0         1
Name: bet, dtype: int64

#### Conclusion: The 515 is the right answer!!!

#### Task 2.

Предположим, что если в признаке win находится пропуск, то выигрыша не было. Здесь два возможных случая:

Человек не делал ставки и ничего не выиграл. То есть просто пришел, посмотрел и ушел.
Человек делал ставку, но не выиграл. Значит, выигрыш на самом деле является отрицательным значением - это проигрыш.
Предлагаем вам написать метод, который заполнит пропуски в признаке win в соответствии с предположением выше. 

Для этого можно применить метод apply() ко всей таблице и передать ему функцию, которая вычисляет размер выигрыша (или проигрыша) по следующей схеме:

- Если значение в столбце win существует (не пропуск) - вернуть это же значение. Это значит, что человек выиграл.
- Если вместо значения в столбце win пропуск, возможны два варианта действий:
- Если в столбце bet в этом же месте стоит 0, вернуть 0. Человек просто зашел посмотреть.
- Если в столбце bet в этом же месте есть ненулевое значение, вернуть то же значение, но с отрицательным знаком. Человек поставил деньги и проиграл их.

На выходе получится столбец без пропусков. Следующим шагом будет замена старого столбца win на новый.

Посчитайте, сколько раз участники ставок проиграли деньги. То есть посчитайте количество строк, для которых в столбце win находится отрицательное значение.

In [9]:
import numpy as np

def fillna_win(row):
#     display(row)
    
    if np.isnan(row['win']):
        result = row['bet'] * -1
    else:
        result = row['win']
    
    return result

In [10]:
new_win = log.apply(lambda row: fillna_win(row), axis=1)
log['win'] = new_win

In [11]:
sum(log['win'] < 0)

347

In [12]:
log['net'] = log['win'] - log['bet']

In [14]:
log['net'].head()

0   -0.0
1   -0.0
2   -0.0
3   -0.0
4   -0.0
Name: net, dtype: float64

In [15]:
sum(log['net']>0)

138

In [16]:
log[log['net']>0]['net'].mean()

80253.33333333333

In [17]:
log[log['net']>0]['net'].median()

5347.0

In [31]:
# прочитаем файл в переменную log
log = pd.read_csv(log_filename, header=None)

# добавим названия колонок user_id, time, bet, win
log.columns = ['user_id','time','bet','win']

In [32]:
log.bet.mean(skipna=False) 

nan

In [33]:
log.bet.mean() 

6785.738144329897

In [34]:
log.bet.sum() / log.bet.dropna().shape[0] 

6785.738144329897

In [35]:
 np.mean(log.bet) 

6785.738144329897

In [36]:
log['bet'].dropna().mean() 

6785.738144329897